### DISCUSSION TITLE
C++ coding convention of the decompiler

### DISCUSSION THREAD

**Discussion by sakiodre** [NationalSecurityAgency/ghidra#5541] [General]
I have been reading the source code of the decompiler for a little over a month now, and too often I need to format it in order to read it.

## First problem: Indentation
Here is a sample block of code taken from [action.cc#L323](https://github.com/NationalSecurityAgency/ghidra/blob/d932173a25bdcb36171f5c0fbaf454ba24f0435d/Ghidra/Features/Decompiler/src/decompile/cpp/action.cc#L323), screenshot to show tabs/spaces
![image](https://github.com/NationalSecurityAgency/ghidra/assets/136492105/1e87ab3b-19bd-4ae1-b5f2-f66a9c84cd07)

Apart from mixing tabs with spaces, they don't even match each other. Not just these specific lines in this particular file, this inconsistency shows up basically everywhere in the code.

## Second problem: Spaces
I guess this kind of formatting is intended as it is consistent across the whole code base, but it is just extremely hard to read and makes a great way to obfuscate the code. 
```cpp
  while((lcount<count)&&((flags&rule_repeatapply)!=0))
  if ((flags&(rule_onceperfunc|rule_oneactperfunc))!=0)
  if ((count>0)||((flags&rule_onceperfunc)!=0))
  for(iter=beginLoc();iter!=endLoc();++iter)
  for(int4 i=0;i<glb->numSpaces();++i)
```
With added spaces, it's now a whole lot easier to read:
```cpp
  while ((lcount < count) && ((flags & rule_repeatapply) != 0))
  if ((flags & (rule_onceperfunc | rule_oneactperfunc)) != 0)
  if ((count > 0) || ((flags & rule_onceperfunc) !=0))
  for (iter = beginLoc(); iter != endLoc(); ++iter)
  for (int4 i = 0; i < glb->numSpaces(); ++i)
```

## Third problem: Function definitions
I wonder what was the motivation for the decision to add a blank line between the function definition and the curly braces (function body)?
```cpp
void PrintC::emitEnumDefinition(const TypeEnum *ct)

{
  // code
}
```
This is not hard to read, by any means, but it is also not the standard at all for any coding convention that I'm aware of. And it's also inconsistent, for example: [printc.cc#L2498](https://github.com/NationalSecurityAgency/ghidra/blob/d932173a25bdcb36171f5c0fbaf454ba24f0435d/Ghidra/Features/Decompiler/src/decompile/cpp/printc.cc#L2498) doesn't have a blank line in between.

## Fourth problem: Using std:: in headers
Take a look at this crazy amount of `using std::` in [error.hh#L38](https://github.com/NationalSecurityAgency/ghidra/blob/d932173a25bdcb36171f5c0fbaf454ba24f0435d/Ghidra/Features/Decompiler/src/decompile/cpp/error.hh#L38).

Although as of now they're all put inside the `ghidra` namespace meaning they're not as much of a problem, but `using std::` in headers is still considered a bad practice, especially in this case where other files depend on this to "using" for them.

## Suggestion
Why don't we follow a standard coding style convention like [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html), and use a code formatter?

**Comment by astrelsky** [comment]
- Regarding the third problem. This is also done in decompiled output 😂.

**Comment by Wall-AF** [comment]
- Regarding "Indentation", tabs appear to have been used to represent 8-spaces, so with an appropriate editor you can see these correctly.  I use Eclipse with Ghidra's source formatting instruction/configuration file!
